driver/controller.go

```diff
@@ -150,6 +165,42 @@ func (s controllerService) CreateVolume(ctx context.Context, req *csi.CreateVolu
 	}, nil
 }
 
+// validateContentSource checks if the request has a data source and returns source volume information.
+func (s controllerService) validateContentSource(ctx context.Context, req *csi.CreateVolumeRequest) (*v1.LogicalVolume, string, error) {
+	volumeSource := req.VolumeContentSource
+
+	if volumeSource == nil {
+		return nil, "", nil
+	}
+
+	switch volumeSource.Type.(type) {
+	case *csi.VolumeContentSource_Snapshot:
+		snapshotID := req.VolumeContentSource.GetSnapshot().GetSnapshotId()
+		if snapshotID == "" {
+			return nil, "", status.Error(codes.NotFound, "Snapshot ID cannot be empty")
+		}
+		snapshotVol, err := s.lvService.GetVolume(ctx, snapshotID)
+		if err != nil {
+			return nil, "", status.Error(codes.NotFound, "failed to find source snapshot")
```

Comment:
is it always notFound?

Reply:
Thanks! Added a generic return for internal errors too :+1: